src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java

```diff
@@ -161,130 +237,766 @@ public ResultSet getTableTypes() throws SQLException {
+     * @param <T> The return type of the Supplier.
+     * @return The return value of the Supplier.
+     * @throws SQLException If a RuntimeException that wraps a SQLException is caught.
+     */
+    private static <T> T liftSQLException(Supplier<T> f) throws SQLException {
+        try {
+            return f.get();
+        } catch (RuntimeException e) {
+            if (e.getCause() instanceof SQLException) {
+                throw (SQLException) e.getCause();
+            }
+            throw e;
+        }
+    }
+
+    // Helper class for representing all info needed to serialize column data for the
+    // getColumns and getColumnPrivileges methods. Intended for use with toGetColumnsDoc
+    // and toGetColumnPrivilegesDoc helpers.
+    private static class GetColumnsDocInfo {
+        String dbName;
+        String tableName;
+        String columnName;
+        MongoJsonSchema columnSchema;
+        boolean isRequired;
+        int idx;
+
+        GetColumnsDocInfo(
+                String dbName,
+                String tableName,
+                String columnName,
+                MongoJsonSchema columnSchema,
+                boolean isRequired,
+                int idx) {
+            this.dbName = dbName;
+            this.tableName = tableName;
+            this.columnName = columnName;
+            this.columnSchema = columnSchema;
+            this.isRequired = isRequired;
+            this.idx = idx;
+        }
+    }
+
+    // Helper for creating BSON documents for the getColumns method. Intended for use
+    // with the getColumnsFromDB helper method which is shared between getColumns and
+    // getColumnPrivileges.
+    private BsonDocument toGetColumnsDoc(GetColumnsDocInfo i) {
+        String bsonType;
+        int nullability;
+        BsonValue dataType;
+        BsonValue decimalDigits = BsonNull.VALUE;
+        BsonValue numPrecRadix;
+        BsonValue charOctetLength = BsonNull.VALUE;
+
+        try {
+            Pair<String, Integer> typeAndNullability =
+                    BsonTypeInfo.getBsonTypeNameAndNullability(i.columnSchema, i.isRequired);
+            bsonType = typeAndNullability.left();
+            nullability = typeAndNullability.right();
+
+            dataType = new BsonInt32(BsonTypeInfo.getJDBCType(bsonType));
+
+            Integer d = BsonTypeInfo.getDecimalDigits(bsonType);
+            if (d != null) {
+                decimalDigits = new BsonInt32(d);
+            }
+
+            numPrecRadix = new BsonInt32(BsonTypeInfo.getNumPrecRadix(bsonType));
+
+            Integer c = BsonTypeInfo.getCharOctetLength(bsonType);
+            if (c != null) {
+                charOctetLength = new BsonInt32(c);
+            }
+
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+
+        BsonValue isNullable =
+                new BsonString(
+                        nullability == columnNoNulls
+                                ? "NO"
```

Comment:
Note : Double-check that SQL-559 has BsonString("Yes) and BSonString("No"). If not, please add it and use for isNullable and isAutoincrement and whereever else the data is "Yes"/"No".

Reply:
Do you mean [SQL-579](https://jira.mongodb.org/browse/SQL-579)?

Reply:
Added a comment calling this out: https://jira.mongodb.org/browse/SQL-579?focusedCommentId=4186564&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-4186564